### PR TITLE
Generalize the authentication backend

### LINKED ISSUE #99: Rework: implement generic OIDCInit and Callback view classes
The view classes must work with specific configuration (sub) classes and respect this configuration.

A POC for this exists in Open Forms, port the code from there.

- [x] Init view
- [x] Authentication backend(s)
- [x] Callback view

### PR DESCRIPTION
**Pull request by sergei-maertens**
Partly closes #99

This PR builds on top of #100, which is responsible for storing which config to apply during the user authentication.

In this PR, that config is actually applied to modify the behaviour/connection based on the specific config class.

In the future, the config classes will be replaced by particular config instances.

The tests have been rewritten to be more decoupled from the implementation details.

### CODE REVIEW

**Review by stevenbal — CHANGES_REQUESTED** [review]
Some minor remarks/questions, looks good overall :+1:

File: mozilla_django_oidc_db/backends.py [hunk 1]
```diff
@@ -1,84 +1,220 @@
...
-    GetAttributeMixin, SoloConfigMixin[T], _OIDCAuthenticationBackend
-):
+class OIDCAuthenticationBackend(BaseBackend):
     """
-    Modifies the default OIDCAuthenticationBackend to use a configurable claim
-    as unique identifier (default `sub`).
+    Custom backend modifying the upstream package behaviour.
+
+    This backend is meant to look up the configuration (class) to use via the config
+    query parameter (pointing to a model class now, probably a unique key in the
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply mozilla_django_oidc_db/backends.py:48]: yep!

File: mozilla_django_oidc_db/backends.py [hunk 2]
```diff
@@ -1,84 +1,220 @@
...
+    OIDC_RP_SIGN_ALGO = dynamic_setting[str](default="HS256")
+    OIDC_RP_IDP_SIGN_KEY = dynamic_setting[str | None](default=None)
+
+    # Custom config/settings from our own models. These map to model fields or
+    # properties on the OpenIDConnectConfigBase model. The settings with defaults can
+    # be provided by subclasses of OpenIDConnectConfigBase or plain old Django settings.
+    OIDCDB_SENSITIVE_CLAIMS = dynamic_setting[Collection[ClaimPath]]()
+    OIDCDB_USERNAME_CLAIM = dynamic_setting[ClaimPath]()
+    OIDCDB_USERINFO_CLAIMS_SOURCE = dynamic_setting[UserInformationClaimsSources](
+        default=UserInformationClaimsSources.userinfo_endpoint
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply mozilla_django_oidc_db/backends.py:81]: they're not specified on the abstract base model, right?

File: mozilla_django_oidc_db/config.py [hunk 3]
```diff
@@ -36,3 +41,100 @@ def get_setting_from_config(config: OpenIDConnectConfigBase, attr: str, *args) -
             return None
         return value_from_config
     return import_from_settings(attr, *args)
+
+
+T = TypeVar("T")
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply mozilla_django_oidc_db/config.py:46]: it's a "placeholder" type for generics! Then you can do things like:

```py
class Foo(Generic[T]):
    bar: T


foo = Foo[Bar]()
reveal_type(foo.bar)  # Bar

foo = Foo[Baz]()
reveal_type(foo.bar)  # Baz
```

File: tests/cassettes/test_init_flow/test_default_config_flow.yaml [hunk 4]
```diff
@@ -1,39 +0,0 @@
-interactions:
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply tests/cassettes/test_init_flow/test_default_config_flow.yaml:1]: no, I switched to `dummy_config` which doesn't use the discovery endpoint to get the real URLs. We don't do live requests with this config (they're not e2e/integration tests with a real keycloak), so we can remove some dead weight :)

File: tests/test_backend.py [hunk 5]
```diff
@@ -1,818 +1,512 @@
...
-    assert backend.get_sensitive_claims_names() == [
-        ["sub"],
-        ["sensitive_claim1"],
-        ["sensitive_claim2"],
-    ]
+    user = backend.authenticate(callback_request)
 
+    # Authentication with the backend should not return a result,
+    # because OIDC is not enabled
+    assert user is None
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply tests/test_backend.py:51]: that's exactly it - the model field has the enabled flag set to `True`, but the `@property` on the custom config ignores that and disables it

File: mozilla_django_oidc_db/backends.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply mozilla_django_oidc_db/backends.py:81]: I double checked, they indeed need to be specified. The reason for this is that the backend expects a `config_class` that's either the class `OpenIDConnectConfigBase` itself, or a subclass of it. Most of the configs here are specified on the subclass `OpenIDConnectConfig`, but custom config classes derive from the abstract base model, not this specific model, so there is no guarantee that the config parameters like `username_claim` will be available, hence, we need default values.

File: mozilla_django_oidc_db/backends.py [hunk 6]
```diff
@@ -126,164 +262,148 @@ def get_userinfo(self, access_token, id_token, payload):
...
 
-    def authenticate(self, *args, **kwargs):
-        if not self.config.enabled:
-            return None
+    @override
+    def filter_users_by_claims(self, claims: JSONObject):
+        """Return all users matching the specified subject."""
+        username = self._extract_username(claims, raise_on_empty=True)
+        if not username:
+            return self.UserModel.objects.none()
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply mozilla_django_oidc_db/backends.py:270]: yep, this is correct! I noticed that test coverage isn't great yet, so I'll be adding some more tests to cover all these code paths.

**Pull request merged**